Contributions/Linux_Memory_Management_Essentials.md

```diff
@@ -0,0 +1,167 @@
+
+[Terms and Abbreviations](#Terms-and-Abbreviations)
+
+[References](#References)
+
+[Disclaimer](#Disclaimer)
+
+[Purpose of the document](#Purpose-of-the-document)
+
+[Structure of the document](#Structure-of-the-document)
+
+[Kernel-space memory allocations](Kernel-space-memory-allocations)
+
+[User-space memory allocations](User-space-memory-allocations)
+
+[License: CC BY-SA 4.0](#License-CC-BY-SA-40)
+
+
+## **Terms and Abbreviations**
+Plese refer to the Linux Kernel documentation.
+   
+
+## **References**
+
+1. [Linux Kernel website](https://www.kernel.org) - <https://www.kernel.org>
+2. ***Interference Scenarios for an ARM64 Linux System***
+3. [CC BY-SA 4.0 Deed | Attribution-ShareAlike 4.0 International | Creative Commons](https://creativecommons.org/licenses/by-sa/4.0/) - <https://creativecommons.org/licenses/by-sa/4.0/> License
+
+
+## **Disclaimer**
+* This document is not intended to be a replacement for understanding the memory management of the Linux,
+  nor it attempts to be an exhaustive analysis of safety implications.
+* Because of the very volatile nature of the code within the Linux Kernel, each of the statements made
+  below should not be taken at face value, but rather verified, for any Linux kernel version following
+  the one used while writing the document.
+* When referring to specific HW features, the document refers to the ARM64 architecture.
+
+## **Purpose of the document**
+This document aims to providing an holistic view of what happens in Linux memory management, so that
+one is at least aware of certain features and can use this document as jumping pad toward more detailed documentation.
+Or even to the code base.
+
+## **Structure of the document**
+The document is divided in two parts, based on hte destinatary of the memory allocations discussed: kernel-space or user-space.
+Individual points are numbered for ease of reference, but the numbering is not meant to represent any sequence.
+
+## **Memory management in Linux**
+
+### **Kernel-space memory allocations**
+
+#### **Facts**
+1. unlike processes memory, kernel memory pages are not swapped, nor dropped silently by the kernel itself,
+   although an hypervisor will do to a VM what the kernel does to a process (but this is beyond the control of the kernel)
+2. the kernel context (usually EL1 on ARM64) uses one single memory map (page tables) across all the cores
+   executing in kernel mode
+3. on 64 bit systems (e.g. ARM64 and x86_64), usually almost all physical memory is mapped in a
+   (semi)contiguous (there can be holes) range. Memory within this range is both virtually and physically contiguous.
+4. physically contiguous memory is treated as a scarce resource, and typically is not provided to userspace,
+   unless it explicitly asks for it (e.g. for implementing some DMA buffer)
+5. the kernel can access userspace in two ways:
+   1. through the userspace mapping
+      1. it is done in exceptional cases, like copy_to_user_copy_from_user
```

Comment:
@dweingaertner : "exceptional cases" was referred to the fact that - codewise (spatially?) - the kernel does not have acces to the memory map of a process, but now I see that it might be misunderstood in a temporal sense, where the kernel indeed accesses  userspace memory all the time, just through a limited set of code paths.
I will have to be more verbose on this.

Reply:
@shetze yes, you are right, I forgot those 2 - I will add them, thanks. WRT the meaning of "exceptional", please read my comments above. But yes, it's obviously not well worded and I will try to describe it better.